fl4health/strategies/fedpca.py

```diff
@@ -0,0 +1,269 @@
+                In this case, the parameters are the new singular vectors and their corresponding singular values.
+        """
+        if not results:
+            return None, {}
+        # Do not aggregate if there are failures and failures are not accepted
+        if not self.accept_failures and failures:
+            return None, {}
+
+        client_singular_values = []
+        client_singular_vectors = []
+        for _, fit_res in results:
+            A = parameters_to_ndarrays(fit_res.parameters)
+            singular_vectors, singular_values = A[0], A[1]
+            client_singular_vectors.append(singular_vectors)
+            client_singular_values.append(singular_values)
+
+        if self.svd_merging:
+            log(INFO, "Performing SVD-based merging.")
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_svd(
+                client_singular_vectors, client_singular_values
+            )
+        else:
+            # use qr merging instead
+            log(INFO, "Performing QR-based merging.")
+            merged_singular_vectors, merged_singular_values = self.merge_subspaces_qr(
+                client_singular_vectors, client_singular_values
+            )
+        parameters = ndarrays_to_parameters([merged_singular_vectors, merged_singular_values])
+
+        # Aggregate custom metrics if aggregation fn was provided
+        metrics_aggregated = {}
+        if self.fit_metrics_aggregation_fn:
+            fit_metrics = [(res.num_examples, res.metrics) for _, res in results]
+            metrics_aggregated = self.fit_metrics_aggregation_fn(fit_metrics)
+        elif server_round == 1:  # Only log this warning once
+            log(WARNING, "No fit_metrics_aggregation_fn provided")
+
+        return parameters, metrics_aggregated
+
+    def merge_subspaces_svd(
+        self, client_singular_vectors: NDArrays, client_singular_values: NDArrays
+    ) -> Tuple[NDArray, NDArray]:
+        """
+        Produce the principal components (PCs) for all the data distributed across clients by merging the PCs
+        belonging to each local dataset.
+
+        Each clients sends a matrix whose columns are the local principal components to the server. The corresponding
+        singular values are also shared.
+
+        The server then arranges the local PCs into a block matrix, then perform SVD.
+
+        For example, if U_i denotes the matrix of PCs and S_i denotes the diagonal matrix of
+        singular values for client i, and there are n clients, then merging is done by
+        performing SVD on the matrix
+
+        B = [U_1 @ S_1 | U_2 @ S_2 | ... | U_n @ S_n],
+
+        where the new (left) singular vectors are returned as the merging result.
+
+        This implementation assumes that the *columns* of U_i are the PCs.
+
+        If
+
+        U @ S @ V.T = B
+
+        is the SVD of B, then it turns out that U = A @ U',
```

Comment:
In the implementation of pca.py, we see to be returning the right singular vectors rather than the left singular vectors. For standard PCA, we can perform projection with either through different computations. However, does this procedure still work?

Reply:
Ah I see the confusion. I should've made this more clear. Here we are returning the left singular vectors because the algorithm assumes that the **columns** of the data matrix (sent by clients to the server) are the local principal components. 

In contrast, `pca.py`, assumes that the **rows** of the data matrix are the data points. Basically the input data matrices are "transposes" of each other in the two cases, so their SVDs are also flipped. That's why we are returning the left singular vectors here and right singular vectors in `pca.py`. 

I did this because the first dimension of a dataset is usually the batch dimension, so it makes sense that the data points would be rows, hence the design in `pca.py`. As for `fedpca.py`, I assumed the local principal components come in columns because all the papers I read on this seem to assume so. 

I agree this is pretty confusing. So I think we can either clarify things better in the comments or change the implementations to be consistent, I'm okay with either approach